language/move-prover/bytecode-to-boogie/src/dataflow_analysis.rs

```diff
@@ -0,0 +1,117 @@
+// Copyright (c) The Libra Core Contributors
+// SPDX-License-Identifier: Apache-2.0
+
+//! Adapted from AbstractInterpreter for Bytecode, this module defines the data-flow analysis
+//! framework for stackless bytecode.
+
+use bytecode_verifier::absint::{AbstractDomain, JoinResult};
+use bytecode_verifier::control_flow_graph::{BlockId, ControlFlowGraph};
+use stackless_bytecode_generator::stackless_bytecode::StacklessBytecode;
+use std::collections::HashMap;
+
+pub struct BlockState<State: Clone> {
+    pub pre: State,
+    pub post: State,
+}
+
+pub type StateMap<State> = HashMap<BlockId, BlockState<State>>;
+
+/// Take a pre-state + instruction and mutate it to produce a post-state。
+pub trait TransferFunctions {
+    type State: AbstractDomain + Clone;
+
+    /// Execute local@instr found at index local@index in the current basic block from pre-state
+    /// local@pre.
+    /// Auxilary data from the analysis that is not part of the abstract state can be collected by
+    /// mutating local@self.
+    /// The last instruction index in the current block is local@last_index. Knowing this
+    /// information allows clients to detect the end of a basic block and special-case appropriately
+    fn execute(
+        &mut self,
+        pre: &mut Self::State,
+        instr: &StacklessBytecode,
+        index: usize,
+        last_index: usize,
+    );
+}
+
+pub trait DataflowAnalysis: TransferFunctions {
+    fn analyze_function(
+        &mut self,
+        initial_state: Self::State,
+        instrs: &[StacklessBytecode],
+        cfg: &dyn ControlFlowGraph,
+    ) -> StateMap<Self::State> {
+        let mut state_map: StateMap<Self::State> = StateMap::new();
+        let entry_block_id = cfg.entry_block_id();
+        let mut work_list = vec![entry_block_id];
+        state_map.insert(
+            entry_block_id,
+            BlockState {
+                pre: initial_state.clone(),
+                post: initial_state.clone(),
+            },
+        );
+
+        while let Some(block_id) = work_list.pop() {
+            let block_res = match state_map.get_mut(&block_id) {
+                Some(res) => res,
+                None => unreachable!("Missing result for block {}", block_id),
+            };
+
+            let mut state = block_res.pre.clone();
+
+            self.execute_block(block_id, &mut state, &instrs, cfg);
```

Comment:
If this is a function, we would just have `let state = self.execute_block(block_id, &block_res.pre, &instrs, cfg);`